cpp/arcticdb/storage/s3/ec2_utils.hpp

```diff
@@ -0,0 +1,72 @@
+#include <aws/core/platform/Environment.h>
+#include <curl/curl.h>
+
+namespace arcticdb::storage::s3 {
+
+// We only care about the response codes, so we just pass a dummy write function to libcurl to not print the responses.
+size_t write_callback([[maybe_unused]] void *buffer, size_t size, size_t nmemb, [[maybe_unused]] void *userp) {
+    return size * nmemb;
+}
+
+// A fast check to identify whether we're running on AWS EC2.
+// According to AWS docs a reliable way to identify if we're in EC2 is to query the instance metadata service:
+// https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/identify_ec2_instances.html
+//
+// Since there are two versions IMDSv1 and IMDSv2 we first try to connect to v2 and if we fail then attempt the legacy
+// v1 connection. If both fail we're most likely running outside of EC2 (unless IMDS is under heavy load and takes more
+// than 100ms to respond)
+bool has_connection_to_ec2_imds(){
+    CURL *curl = curl_easy_init();

```

Comment:
Might be more readable to shortcircuit, `if (!curl) // return false immediately` and then the main implementation doesn't need to be nested.